instrumentation/nats/nats-2.17/library/src/main/java/io/opentelemetry/instrumentation/nats/v2_17/OpenTelemetryConnection.java

```diff
@@ -0,0 +1,328 @@
+  // Duration timeout);
+  // CompletableFuture<Message> request(Message message);
+  // CompletableFuture<Message> requestWithTimeout(Message message, Duration timeout);
+  @SuppressWarnings("unchecked")
+  private CompletableFuture<Message> requestAsync(Method method, Object[] args) throws Throwable {
+    String subject = null;
+    Headers headers = null;
+    byte[] body = null;
+    Duration timeout = null;
+
+    if ((method.getParameterCount() == 2)
+        && method.getParameterTypes()[0] == String.class
+        && method.getParameterTypes()[1] == byte[].class) {
+      subject = (String) args[0];
+      body = (byte[]) args[1];
+    } else if ((method.getParameterCount() == 3)
+        && method.getParameterTypes()[0] == String.class
+        && method.getParameterTypes()[1] == byte[].class) {
+      subject = (String) args[0];
+      body = (byte[]) args[1];
+      timeout = (Duration) args[2];
+    } else if ((method.getParameterCount() == 3)
+        && method.getParameterTypes()[0] == String.class
+        && method.getParameterTypes()[1] == Headers.class
+        && method.getParameterTypes()[2] == byte[].class) {
+      subject = (String) args[0];
+      headers = (Headers) args[1];
+      body = (byte[]) args[2];
+    } else if ((method.getParameterCount() == 4)
+        && method.getParameterTypes()[0] == String.class
+        && method.getParameterTypes()[1] == Headers.class
+        && method.getParameterTypes()[2] == byte[].class) {
+      subject = (String) args[0];
+      headers = (Headers) args[1];
+      body = (byte[]) args[2];
+      timeout = (Duration) args[3];
+    } else if ((method.getParameterCount() == 1)
+        && method.getParameterTypes()[0] == Message.class) {
+      subject = ((Message) args[0]).getSubject();
+      headers = ((Message) args[0]).getHeaders();
+      body = ((Message) args[0]).getData();
+    } else if ((method.getParameterCount() == 2)
+        && method.getParameterTypes()[0] == Message.class) {
+      subject = ((Message) args[0]).getSubject();
+      headers = ((Message) args[0]).getHeaders();
+      body = ((Message) args[0]).getData();
+      timeout = (Duration) args[1];
+    }
+
+    Context parentContext = Context.current();
+    headers = NatsMessageWritableHeaders.create(headers);
+    NatsRequest natsRequest = null;
+
+    if (subject != null) {
+      natsRequest = NatsRequest.create(delegate, subject, null, headers, body);
+    }
+
+    if (natsRequest == null || !producerInstrumenter.shouldStart(parentContext, natsRequest)) {
+      return (CompletableFuture<Message>) invokeMethod(method, delegate, args);
+    }
+
+    NatsRequest notNullNatsRequest = natsRequest;
+    Context context = producerInstrumenter.start(parentContext, notNullNatsRequest);
+
+    CompletableFuture<Message> future;
+    if (timeout != null) {
+      future = delegate.requestWithTimeout(subject, headers, body, timeout);
```

Comment:
@laurit checking the implementation, this can throw outside of the future. Should I surround also with tryCatch in order to properly propagate the error at the otel level?

<img width="775" height="94" alt="Screenshot 2025-09-04 at 17 41 59" src="https://github.com/user-attachments/assets/64cba9df-8bed-4607-aa05-5c068ee96afe" />
<img width="478" height="202" alt="Screenshot 2025-09-04 at 17 42 12" src="https://github.com/user-attachments/assets/561d7099-8ec0-4ab4-9a8f-021f6064b29b" />

Reply:
what happens also, in this case? The future might not be started, so the 
```java
    return future.whenComplete(
        (result, exception) -> {
          if (result != null) {
            NatsRequest response = NatsRequest.create(delegate, result);
            producerInstrumenter.end(context, notNullNatsRequest, response, exception);
          } else {
            producerInstrumenter.end(context, notNullNatsRequest, null, exception);
          }
        });
```
will never be reached?